src/setup_payload/AdditionalDataPayloadGenerator.cpp

```diff
@@ -51,35 +51,34 @@ AdditionalDataPayloadGenerator::generateAdditionalDataPayload(uint16_t lifetimeC
 {
     System::PacketBufferTLVWriter writer;
     TLVWriter innerWriter;
-    char rotatingDeviceIdBuffer[RotatingDeviceId::kHexMaxLength];
+    uint8_t rotatingDeviceIdBuffer[RotatingDeviceId::kHexMaxLength];
     size_t rotatingDeviceIdBufferSize = 0;
 
     // Initialize TLVWriter
     writer.Init(chip::System::PacketBufferHandle::New(chip::System::PacketBuffer::kMaxSize));
 
-    ReturnErrorOnFailure(writer.OpenContainer(AnonymousTag, kTLVType_Structure, innerWriter));
+    ReturnErrorOnFailure(writer.OpenContainer(RotatingDeviceId::DiscoveryExtensionDescriptorTag, kTLVType_Structure, innerWriter));
 
     if (additionalDataFields.Has(AdditionalDataFields::RotatingDeviceId))
     {
         // Generating Device Rotating Id
-        ReturnErrorOnFailure(generateRotatingDeviceId(lifetimeCounter, serialNumberBuffer, serialNumberBufferSize,
-                                                      rotatingDeviceIdBuffer, ArraySize(rotatingDeviceIdBuffer),
-                                                      rotatingDeviceIdBufferSize));
+        ReturnErrorOnFailure(generateRotatingDeviceIdInBinary(lifetimeCounter, serialNumberBuffer, serialNumberBufferSize,
+                                                              rotatingDeviceIdBuffer, ArraySize(rotatingDeviceIdBuffer),
+                                                              rotatingDeviceIdBufferSize));
 
         // Adding the rotating device id to the TLV data
-        ReturnErrorOnFailure(innerWriter.PutString(ContextTag(kRotatingDeviceIdTag), rotatingDeviceIdBuffer,
-                                                   static_cast<uint32_t>(rotatingDeviceIdBufferSize)));
+        ReturnErrorOnFailure(innerWriter.PutBytes(ContextTag(kRotatingDeviceIdTag), rotatingDeviceIdBuffer,
+                                                  static_cast<uint32_t>(rotatingDeviceIdBufferSize)));
     }
 
     ReturnErrorOnFailure(writer.CloseContainer(innerWriter));
 
     return writer.Finalize(&bufferHandle);
 }
 
-CHIP_ERROR AdditionalDataPayloadGenerator::generateRotatingDeviceId(uint16_t lifetimeCounter, const char * serialNumberBuffer,
-                                                                    size_t serialNumberBufferSize, char rotatingDeviceIdBuffer[],
-                                                                    size_t rotatingDeviceIdBufferSize,
-                                                                    size_t & rotatingDeviceIdValueOutputSize)
+CHIP_ERROR AdditionalDataPayloadGenerator::generateRotatingDeviceIdInBinary(
+    uint16_t lifetimeCounter, const char * serialNumberBuffer, size_t serialNumberBufferSize, uint8_t * rotatingDeviceIdBuffer,
```

Comment:
Ideally this API would use spans for better type safety... Followup, please.

Reply:
sure, i can send a follow up PR for these minor updates